In [3]:
#imports
import numpy as np
import numpy.random as rnd
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import os

222240301

## Parte 1: Limpieza y transformación

### (2 pts) 1.1 Carga de datos

1. Lea el archivo `pib2020-2025.csv` en un DataFrame llamado `gdpData`.  
   - Muestre sus primeras filas y el tipo de cada columna.
2. Lea el archivo `WHO_life_expectancy.csv` en un DataFrame llamado `lifeData`.  
   - Muestre sus primeras filas y el tipo de cada columna.

Comente brevemente **qué representa cada DataFrame** y cuál crees que será el “link” natural entre ellos.

---

In [17]:
path1 = os.path.join('data_2025', 'pib2020-2025.csv')
path2 = os.path.join('data_2025', 'WHO_life_expectancy.csv')

pib2 = pd.read_csv(path1)
gdpData = pd.read_csv(path1)
LifeData = pd.read_csv(path2)

display(gdpData)
display(LifeData)
def infodf(df):
    print('INFO')
    print(df.info())
    print('')
    print('COLUMNAS')
    print(df.columns)
    print('')
    print('MEDIDAS ESTADISTICAS')
    print(df.describe())

infodf(gdpData)
infodf(LifeData)

    
    

,Country,2020,2021,2022,2023,2024,2025
0,Afghanistan,20136,14278.0,14501.0,17248.0,NaN,NaN
1,Albania,15271,18086.0,19185.0,23388.0,27259.0,28372.0
2,Algeria,164774,185850.0,225709.0,247789.0,264913.0,268885.0
3,Andorra,2885,3325.0,3376.0,3786.0,4038.0,4035.0
4,Angola,66521,84375.0,142442.0,109764.0,115946.0,113343.0
...,...,...,...,...,...,...,...
191,Vietnam,346310,370076.0,411068.0,433008.0,459472.0,490970.0
192,Palestine,15532,18109.0,19166.0,17848.0,NaN,NaN
193,Yemen,20220,19394.0,23534.0,19412.0,19101.0,17401.0
194,Zambia,18138,22096.0,29164.0,27578.0,26326.0,28910.0


,IndicatorCode,Indicator,ParentLocationCode,ParentLocation,Location type,SpatialDimValueCode,Location,Period type,Period,Dim1 type,Dim1,Value
0,WHOSIS_000001,Life expectancy at birth (years),AFR,Africa,Country,LSO,Lesotho,Year,2021,Sex,Male,48.7 [47.8-49.7]
1,WHOSIS_000001,Life expectancy at birth (years),AFR,Africa,Country,CAF,Central African Republic,Year,2021,Sex,Male,49.6 [48.5-50.9]
2,WHOSIS_000001,Life expectancy at birth (years),AFR,Africa,Country,LSO,Lesotho,Year,2021,Sex,Both sexes,51.5 [50.5-52.6]
3,WHOSIS_000001,Life expectancy at birth (years),AFR,Africa,Country,SWZ,Eswatini,Year,2021,Sex,Male,51.6 [50.7-52.8]
4,WHOSIS_000001,Life expectancy at birth (years),EMR,Eastern Mediterranean,Country,SOM,Somalia,Year,2021,Sex,Male,51.7 [50.6-53.1]
...,...,...,...,...,...,...,...,...,...,...,...,...
24415,WHOSIS_000015,Life expectancy at age 60 (years),EUR,Europe,Country,CHE,Switzerland,Year,2000,Sex,Female,24.8 [24.8-24.9]
24416,WHOSIS_000015,Life expectancy at age 60 (years),WPR,Western Pacific,Country,AUS,Australia,Year,2000,Sex,Female,24.9 [24.8-24.9]
24417,WHOSIS_000015,Life expectancy at age 60 (years),EUR,Europe,Country,FRA,France,Year,2000,Sex,Female,25.3 [25.2-25.3]
24418,WHOSIS_000015,Life expectancy at age 60 (years),AMR,Americas,Country,NIC,Nicaragua,Year,2000,Sex,Female,25.6 [25.5-25.7]


INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 196 entries, 0 to 195
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Country  196 non-null    object 
 1   2020     196 non-null    int64  
 2   2021     194 non-null    float64
 3   2022     194 non-null    float64
 4   2023     194 non-null    float64
 5   2024     192 non-null    float64
 6   2025     189 non-null    float64
dtypes: float64(5), int64(1), object(1)
memory usage: 10.8+ KB
None

COLUMNAS
Index(['Country', '2020', '2021', '2022', '2023', '2024', '2025'], dtype='object')

MEDIDAS ESTADISTICAS
               2020          2021          2022          2023          2024  \
count  1.960000e+02  1.940000e+02  1.940000e+02  1.940000e+02  1.920000e+02   
mean   4.378886e+05  5.043504e+05  5.255064e+05  5.486173e+05  5.756871e+05   
std    1.942936e+06  2.220864e+06  2.347886e+06  2.453040e+06  2.573189e+06   
min    5.200000e+01  6.200000e+01  6.100000e+01  6

### (5 pts) 1.2 Limpieza básica de `gdpData`

Sobre `gdpData`:

1. Revise cuántos valores faltantes (`NaN`) hay por columna.  
2. Elimine filas donde **todas** las columnas de años (2020–2025) sean `NaN`.  
3. Para los valores faltantes que queden:
   - Use interpolación **por fila (por país)** a lo largo de los años (por ejemplo, usando `interpolate(axis=1)`),  
     y si aún quedan `NaN` en los extremos, reemplácelos por el valor más cercano disponible en el tiempo.
4. Transforme el DataFrame a formato **largo** (tidy) usando `melt`, de manera que quede con columnas:
   - `Country`, `Year` (int), `GDP` (float).
5. Asegúrese de que `Year` sea entero y `GDP` sea numérico.

Comente brevemente qué ventajas tiene el formato largo para el análisis.

---

In [18]:
print(gdpData.isna().sum())
year_cols = ["2020", "2021", "2022", "2023", "2024", "2025"]

# Eliminar filas donde TODOS los años son NaN
gdpData = gdpData.dropna(subset=year_cols, how="all")
# Asegurar que las columnas de años sean numéricas (por si vienen como str)
gdpData[year_cols] = gdpData[year_cols].apply(pd.to_numeric, errors="coerce")

# Interpolar por FILA (axis=1), a lo largo de los años
gdpData[year_cols] = gdpData[year_cols].interpolate(axis=1)

# Si quedan NaN en los extremos, rellenar usando el valor más cercano en el tiempo
# Primero hacia la derecha (bfill), luego hacia la izquierda (ffill), todo por fila
gdpData[year_cols] = gdpData[year_cols].bfill(axis=1).ffill(axis=1)
gdp_long = gdpData.melt(
    id_vars="Country",          # columna que se mantiene fija
    value_vars=year_cols,       # columnas que “se derretirán”
    var_name="Year",            # nuevo nombre para la columna de años
    value_name="GDP"            # nuevo nombre para los valores
)

gdp_long["Year"] = gdp_long["Year"].astype(int)
gdp_long["GDP"] = pd.to_numeric(gdp_long["GDP"], errors="coerce")
gdp_long = gdp_long.dropna(subset=["GDP"])
display(gdp_long)

Country    0
2020       0
2021       2
2022       2
2023       2
2024       4
2025       7
dtype: int64


,Country,Year,GDP
0,Afghanistan,2020,20136.0
1,Albania,2020,15271.0
2,Algeria,2020,164774.0
3,Andorra,2020,2885.0
4,Angola,2020,66521.0
...,...,...,...
1171,Vietnam,2025,490970.0
1172,Palestine,2025,17848.0
1173,Yemen,2025,17401.0
1174,Zambia,2025,28910.0


### Preparacion dataframe pib 2
Caso solo rellenar los valores NaN con la media de años anteriores 

In [21]:
caso1 = "media"  
# caso1 = "cero"


year_cols = ["2020", "2021", "2022", "2023", "2024", "2025"]

# 1. NaN por columna
print("NaN por columna al inicio:")
print(pib2.isna().sum())

pib2 = pib2.dropna(subset=year_cols, how="all")

pib2[year_cols] = pib2[year_cols].apply(pd.to_numeric, errors="coerce")

if caso1 == "media":
    def fill_prev_mean(row):
        prev_vals = []
        for col in year_cols:
            if pd.isna(row[col]):
                if prev_vals:
                    row[col] = np.mean(prev_vals)
            else:
                prev_vals.append(row[col])
        return row

    pib2[year_cols] = pib2[year_cols].apply(fill_prev_mean, axis=1)
    
elif caso1 == "cero":
    pib2[year_cols] = [pib2year_cols].fillna(0)


# 4. Pasar a formato largo (tidy)
#gdp_long = gdpData.melt(
#   id_vars="Country",
#   value_vars=year_cols,
#   var_name="Year",
#   value_name="GDP"
#)

#gdp_long["Year"] = gdp_long["Year"].astype(int)
#gdp_long["GDP"] = pd.to_numeric(gdp_long["GDP"], errors="coerce")

print("\nPrimeras filas del resultado:")
display(pib2)

NaN por columna al inicio:
Country    0
2020       0
2021       0
2022       0
2023       0
2024       0
2025       0
dtype: int64

Primeras filas del resultado:


,Country,2020,2021,2022,2023,2024,2025
0,Afghanistan,20136.0,14278.0,14501.0,17248.0,16540.75,16540.75
1,Albania,15271.0,18086.0,19185.0,23388.0,27259.00,28372.00
2,Algeria,164774.0,185850.0,225709.0,247789.0,264913.00,268885.00
3,Andorra,2885.0,3325.0,3376.0,3786.0,4038.00,4035.00
4,Angola,66521.0,84375.0,142442.0,109764.0,115946.00,113343.00
...,...,...,...,...,...,...,...
191,Vietnam,346310.0,370076.0,411068.0,433008.0,459472.00,490970.00
192,Palestine,15532.0,18109.0,19166.0,17848.0,17663.75,17663.75
193,Yemen,20220.0,19394.0,23534.0,19412.0,19101.00,17401.00
194,Zambia,18138.0,22096.0,29164.0,27578.0,26326.00,28910.00


### (6 pts) 1.3 Preparación de `lifeData`

En el DataFrame `lifeData`:

1. Filtre solo las filas que correspondan a:
   - Indicador **“Life expectancy at birth (years)”** (`IndicatorCode == 'WHOSIS_000001'`).
   - Dimensión de sexo **“Both sexes”** (`Dim1 == 'Both sexes'`).
   - Períodos entre **2015 y 2025** (ambos incluidos).
2. La columna `Value` tiene un string del tipo `"48.7 [47.8-49.7]"`.  
   - Cree una nueva columna `LifeExpectancy` que tenga solamente el número central (ej: 48.7) como `float`.  
     (Pista: puede separar por espacio, tomar la primera parte, o usar operaciones de string).
3. Deje solo las columnas relevantes:
   - `Location` (nombre del país),  
   - `ParentLocation` (región OMS),  
   - `Period` (año),  
   - `LifeExpectancy`.

Renombre las columnas para que queden como:

- `Country`, `Region`, `Year`, `LifeExpectancy`.

---

### (4 pts) 1.4 DataFrame combinado `worldHealth`

Ahora queremos un DataFrame que combine PIB y expectativa de vida por país y año.

1. Haga un `merge` entre `gdpData` (en formato largo) y el DataFrame de la pregunta 1.3 usando:
   - `Country` y `Year` como llaves.
2. Guarde el resultado en un DataFrame llamado `worldHealth`, con las columnas:
   - `Country`, `Region`, `Year`, `GDP`, `LifeExpectancy`.
3. Revise cuántas filas tiene `worldHealth` y si quedan valores faltantes en `GDP` o `LifeExpectancy`.  
   - Si lo estima razonable, elimine filas con datos faltantes (explique su decisión en 1–2 líneas).

---

### (EXTRA 1.A – práctica) Limpieza extra por consistencia

- Verifique si hay países que aparecen con nombres distintos entre ambos datasets (ej: “United States” vs “United States of America”).  
  En caso de encontrarlos, corrija estos nombres en el código (por ejemplo, usando `replace`) para mejorar el `merge`.  
- Vuelva a construir `worldHealth` y compare cuántas filas adicionales gana.

---


## Parte 2: EDA (Análisis Exploratorio)

Usaremos el DataFrame `worldHealth` construido en la Parte 1.

### (4 pts) 2.1 Resumen global por año

1. Para cada `Year` en `worldHealth`:
   - Calcule el **PIB promedio**, **PIB mediano**,  
     y la **expectativa de vida promedio** y **mediana**.
2. Presente estos resultados en un DataFrame o tabla ordenada por año.
3. Comente brevemente **cómo ha cambiado** (según estos datos) el PIB promedio mundial y la expectativa de vida promedio entre 2020 y 2025.

---

### (4 pts) 2.2 Distribuciones por región

1. Elija un año de interés, por ejemplo **2021**.  
   Filtre `worldHealth` solo a ese año.
2. Para ese año:
   - Haga un boxplot de **LifeExpectancy** por `Region`.
   - Haga un boxplot de **GDP** por `Region`.
3. Comente:
   - ¿Qué regiones parecen tener, en general, mayor expectativa de vida?  
   - ¿Coincide esto con las regiones de mayor PIB?

---

### (5 pts) 2.3 Deciles de PIB y brecha de expectativa de vida

En un año fijo (por ejemplo **2023**):

1. Calcule el **decil de PIB** de cada país (1 = PIB más bajo, 10 = PIB más alto), usando `pd.qcut` o técnica similar.
2. Para los países en el decil 1 (PIB más bajo) y decil 10 (PIB más alto), calcule:
   - Promedio y desviación estándar de `LifeExpectancy` en cada grupo.
3. Calcule la diferencia entre la expectativa de vida promedio de ambos grupos.
4. Comente brevemente los resultados:  
   ¿hay una brecha clara entre países de muy bajo y muy alto PIB?

---

### (5 pts) 2.4 Tendencias temporales por región

1. Para cada `Region` y cada `Year`, calcule:
   - PIB promedio,
   - Expectativa de vida promedio.
2. Elija **2 o 3 regiones** (por ejemplo: `Americas`, `Europe`, `Africa`) y, en dos gráficos separados:
   - Grafique la evolución del **PIB promedio** (eje y) vs **Year** (eje x).
   - Grafique la evolución de la **LifeExpectancy promedio** (eje y) vs **Year**.
3. Comente:
   - ¿Alguna región muestra una tendencia clara de mejora o estancamiento en PIB?  
   - ¿Y en expectativa de vida?  
   - ¿Se ve relación visual entre las curvas de PIB y de expectativa de vida?

---

### (EXTRA 2.A – práctica) Brecha de género

A partir de `lifeData` original (no filtrado en 1.3):

1. Construya un DataFrame con columnas:
   - `Country`, `Year`, `LifeExpectancyMale`, `LifeExpectancyFemale`.
   (Use el mismo truco para extraer el valor numérico de `Value`).
2. Calcule la brecha `gap = LifeExpectancyFemale - LifeExpectancyMale` por país y año.
3. Estudie, para un año dado, en qué regiones la brecha es mayor.

---

## Parte 3: Aprendiendo relaciones (Modelos)

### (6 pts) 3.1 Correlación PIB – expectativa de vida

Seleccione un año fijo, por ejemplo **2021**, y trabaje con `worldHealth_2021`.

1. Grafique un **scatter plot** de `LifeExpectancy` vs `GDP`.  
   (Puedes usar el PIB tal cual o transformado con `log(GDP)`; decide y justifica en 1 línea).
2. Calcule el coeficiente de correlación de Pearson entre PIB y expectativa de vida.
3. Comente si la relación parece:
   - débil, moderada o fuerte;
   - y si es positiva o negativa.

---

### (7 pts) 3.2 Regresión simple: PIB → expectativa de vida

Siguiendo con el mismo año (ej: 2021):

1. Separe los datos en entrenamiento y prueba (por ejemplo, 70% train / 30% test, usando `train_test_split`).
2. Entrene un modelo de **regresión lineal** donde la variable independiente sea:
   - `GDP` **o** `log(GDP)` (elige y justifica),  
   y la dependiente sea `LifeExpectancy`.
3. Calcule en el conjunto de prueba:
   - RMSE (raíz error cuadrático medio),
   - R².
4. Interprete:
   - ¿Qué tan bien explica el PIB la variación de la expectativa de vida entre países ese año?

---

### (7 pts) 3.3 Regresión múltiple con año y región

Ahora queremos aprovechar más variables:

1. Usando todo `worldHealth`, defina:
   - `X` con columnas: `GDP`, `Year` (como número) y **dummies** para `Region` (one-hot encoding).
   - `y` como `LifeExpectancy`.
2. Divida en entrenamiento y prueba (70/30 o similar).
3. Entrene un modelo de **regresión lineal múltiple**.
4. Calcule RMSE y R² en entrenamiento y prueba.
5. Comente:
   - ¿Mejora la capacidad de predicción al incluir `Year` y `Region` respecto al modelo simple de 3.2?  
   - ¿Podría haber riesgo de sobreajuste? Justifique brevemente.

---

### (5 pts) 3.4 Clasificación simple: países con alta y baja expectativa de vida

Para un año fijo (por ejemplo 2023):

1. Cree una variable categórica `HighLife` tal que:
   - `1` si `LifeExpectancy` es mayor o igual al cuartil 3 (top 25%),
   - `0` si es menor.
2. Entrene un clasificador simple, por ejemplo **Regresión Logística** o **KNN**, usando:
   - Como features: `GDP` (o `log(GDP)`) y dummies de `Region`.
3. Evalúe el modelo en un conjunto de prueba con:
   - Accuracy,
   - Matriz de confusión (o al menos TP, FP, TN, FN).
4. Comente:
   - ¿Qué tan bien logra este modelo distinguir países de alta vs baja expectativa de vida?  
   - ¿Hay algún tipo de error (falsos positivos / falsos negativos) que te preocuparía más?

---


### (EXTRA 3.A – práctica) Regresión polinomial en log(PIB)

Como ejercicio adicional, para un año específico:

1. Modele `LifeExpectancy` en función de `log(GDP)` usando regresión polinomial de grados 1, 2, 3, …, 6.
2. Para cada grado:
   - Entrene el modelo (usando `PolynomialFeatures` + `LinearRegression`),
   - Calcule RMSE de entrenamiento y prueba.
3. Grafique RMSE versus grado del polinomio y comente:
   - ¿En qué grado parece haber sobreajuste?
   - ¿Qué grado elegirías y por qué?
